crates/iota-indexer/src/backfill/runner.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    cmp,
+    collections::BTreeSet,
+    ops::RangeInclusive,
+    sync::{
+        Arc,
+        atomic::{AtomicUsize, Ordering},
+    },
+    time::Instant,
+};
+
+use futures::{StreamExt, TryStreamExt, stream::unfold};
+use tokio::sync::Mutex;
+use tokio_stream::Stream;
+use tracing::{error, info};
+
+use crate::{
+    backfill::{BackfillTaskKind, get_backfill_task, task::BackfillTask},
+    config::BackfillConfig,
+    db::ConnectionPool,
+    errors::IndexerError,
+};
+
+/// Entry point for orchestrating backfills.
+///
+/// `BackfillRunner` selects the appropriate backfill implementation, splits the
+/// requested checkpoint range into manageable chunks, and dispatches them in
+/// parallel.
+pub struct BackfillRunner;
+
+impl BackfillRunner {
+    /// Execute a backfill over `total_range` using the specified task kind.
+    pub async fn run(
+        runner_kind: BackfillTaskKind,
+        pool: ConnectionPool,
+        backfill_config: BackfillConfig,
+        total_range: RangeInclusive<usize>,
+    ) -> Result<(), IndexerError> {
+        let task = get_backfill_task(runner_kind, *total_range.start()).await;
+        Self::run_impl(pool, backfill_config, total_range, task).await
+    }
+
+    async fn run_impl(
+        pool: ConnectionPool,
+        config: BackfillConfig,
+        total_range: RangeInclusive<usize>,
+        task: Arc<dyn BackfillTask>,
+    ) -> Result<(), IndexerError> {
+        let timer = Instant::now();
+        let processed_counter = Arc::new(AtomicUsize::new(0));
+        // Keeps track of the checkpoint ranges (using starting checkpoint number)
```

Comment:
Here we don't do only checkpoint ranges but others aswell

```suggestion
        // Keeps track of the ranges (using starting range number)
```